File: apps/desktop/src/lib/terminal.ts
import { invokeCommand, on } from "../transport/transport";

export interface TerminalSession {
  /** Send keystrokes / pasted input to the shell's stdin. */
  send: (data: string) => void;
  /** Resize the PTY to match the xterm viewport. */
  resize: (cols: number, rows: number) => void;
  /** Close the session and unsubscribe. */
  close: () => void;
}

/**
 * Open a local shell scoped to `context` (kubectl targets it by default).
 * Runs on the user's machine — distinct from in-pod exec. `onData` receives
 * stdout chunks; `onExit` fires when the shell ends.
 */
export async function startLocalTerminal(
  context: string,
  extraKubeconfigs: string[],
  onData: (chunk: string) => void,
  onExit: () => void,
  size?: { cols: number; rows: number },
): Promise<TerminalSession> {
  const session = await invokeCommand<number>("start_terminal", {
    context,
    extraKubeconfigs,
    cols: size?.cols ?? null,
    rows: size?.rows ?? null,
  });
  const disposeOut = on(`term:out:${session}`, (p) => onData(p as string));
  const disposeExit = on(`term:exit:${session}`, () => onExit());
  return {
    send: (data) => void invokeCommand("terminal_input", { session, data }),
    resize: (cols, rows) => void invokeCommand("terminal_resize", { session, cols, rows }),
    close: () => {
      disposeOut();
      disposeExit();
      void invokeCommand("terminal_close", { session });
    },
  };
}

File: apps/desktop/src-tauri/src/terminal.rs
//! Tauri adapter for the local kubectl terminal: the PTY core lives in
//! srelens_streams::terminal; this module maps the Tauri command surface and
//! merges desktop kubeconfig discovery with any pasted/extra kubeconfigs.

use std::sync::Arc;

use srelens_streams::terminal::TerminalManager;
use tauri::{AppHandle, State};

use crate::sink::TauriSink;

/// Start a local shell scoped to `context`. Returns the session id; output
/// streams on `term:out:<channel>` and a `term:exit:<channel>` event fires
/// when it ends, where `channel` is the caller-supplied subscription token.
#[tauri::command]
pub async fn start_terminal(
    context: String,
    extra_kubeconfigs: Vec<String>,
    channel: String,
    cols: Option<u16>,
    rows: Option<u16>,
    app: AppHandle,
    manager: State<'_, TerminalManager>,
) -> Result<u64, String> {
    let mut paths = crate::capabilities::default_kubeconfig_paths();
    paths.extend(extra_kubeconfigs.iter().map(std::path::PathBuf::from));
    manager
        .start(
            Arc::new(TauriSink(app)),
            context,
            paths,
            channel,
            cols,
            rows,
        )
        .await
}

/// Forward keystrokes / pasted input to a terminal's stdin.
#[tauri::command]
pub async fn terminal_input(
    session: u64,
    data: String,
    manager: State<'_, TerminalManager>,
) -> Result<(), String> {
    manager.input(session, &data);
    Ok(())
}

/// Resize a terminal's PTY (columns/rows) to match the xterm viewport.
#[tauri::command]
pub async fn terminal_resize(
    session: u64,
    cols: u16,
    rows: u16,
    manager: State<'_, TerminalManager>,
) -> Result<(), String> {
    manager.resize(session, cols, rows);
    Ok(())
}

/// Close a terminal: kill the shell and drop the session.
#[tauri::command]
pub async fn terminal_close(
    session: u64,
    manager: State<'_, TerminalManager>,
) -> Result<(), String> {
    manager.close(session);
    Ok(())
}
